Toggl.iOS/ViewSources/DateRangePickerViewSource.cs

```diff
@@ -0,0 +1,95 @@
+﻿using Foundation;
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Reactive.Linq;
+using System.Reactive.Subjects;
+using Toggl.Core.UI.Parameters;
+using Toggl.Core.UI.ViewModels.DateRangePicker;
+using Toggl.Core.UI.ViewModels.ReportsCalendar;
+using Toggl.iOS.Views;
+using UIKit;
+
+namespace Toggl.iOS.ViewSources
+{
+    public sealed class DateRangePickerViewSource : UICollectionViewSource
+    {
+        private const string cellIdentifier = nameof(ReportsCalendarViewCell);
+
+        private readonly UICollectionView collectionView;
+        private readonly ISubject<DateRangePickerMonthInfo> currentPageNotScrollingSubject = new Subject<DateRangePickerMonthInfo>();
+        private readonly ISubject<DateTime> dayTaps = new Subject<DateTime>();
+
+        private IImmutableList<DateRangePickerMonthInfo> months = ImmutableList<DateRangePickerMonthInfo>.Empty;
+        private int? page;
+
+        public IObservable<DateTime> DayTaps { get; }
+        public IObservable<DateRangePickerMonthInfo> CurrentPageNotScrollingObservable { get; }
+
+        public DateRangePickerViewSource(UICollectionView collectionView)
+        {
+            this.collectionView = collectionView;
+
+            collectionView.RegisterNibForCell(ReportsCalendarViewCell.Nib, cellIdentifier);
+
+            DayTaps = dayTaps.AsObservable();
+            CurrentPageNotScrollingObservable = currentPageNotScrollingSubject.AsObservable().DistinctUntilChanged();
+        }
+
+        public override UICollectionViewCell GetCell(UICollectionView collectionView, NSIndexPath indexPath)
+        {
+            var month = months[indexPath.Section];
+            var date = month.DisplayDates.Beginning.AddDays(indexPath.Row);
+            var isToday = month.Today == date;
+            var isSelected = month.Selection?.Contains(date) ?? false;
+            var isInCurrentMonth = date.Month == month.Month;
+            var isSelectionStart = month.IsDateTheFirstSelectedDate(date);
+            var isSelectionEnd = month.IsDateTheLastSelectedDate(date);
+
+            var cell = collectionView.DequeueReusableCell(cellIdentifier, indexPath) as ReportsCalendarViewCell;
+
+            cell.Item = (date, isToday, isSelected, isInCurrentMonth, isSelectionStart, isSelectionEnd);
+            return cell;
+        }
+
+        public void UpdateMonths(IImmutableList<DateRangePickerMonthInfo> newMonths)
+        {
+            months = newMonths;
+            collectionView.ReloadData();
+        }
+
+        public void ScrollToCurrentPage()
+        {
+            if (page == null)
+            {
+                for (var i = 0; i < months.Count; i++)
+                {
+                    if (months[i].IsSelectionBeginningBoundary)
+                        page = i;
+                }
+            }
+
+            if (page == null)
```

Comment:
With the above change, you'd switch this into

```suggestion
            if (page < 0)
```

as the `FindIndex` returns `-1` for not-found.